python/dgl/nn/pytorch/graph_transformer.py

```diff
@@ -336,3 +332,186 @@ def forward(self, g, edge_feat):
+        self.embedding_table = nn.Embedding(
+            max_dist + 1, num_heads, padding_idx=0
+        )
+
+    def forward(self, g):
+        """
+        Parameters
+        ----------
+        g : DGLGraph
+            A DGLGraph to be encoded, which must be a homogeneous one.
+
+        Returns
+        -------
+        torch.Tensor
+            Return attention bias as spatial encoding of shape
+            :math:`(batch_size, N, N, num_heads)`,
+            where :math:`N` is the maximum number of nodes
+            and batch_size is the batch size of the input graph.
+        """
+        device = g.device
+        g_list = unbatch(g)
+        max_num_nodes = th.max(g.batch_num_nodes())
+        spatial_encoding = []
+
+        for ubg in g_list:
+            num_nodes = ubg.num_nodes()
+            dist = th.clamp(
+                shortest_dist(ubg, root=None, return_paths=False),
+                min=0, max=self.max_dist
+            )
+            # shape: [n, n, h], n = num_nodes, h = num_heads
+            dist_embedding = self.embedding_table(dist)
+            # [n, n, h] -> [N, N, h], N = max_num_nodes, padded with -inf
+            padded_encoding = th.full(
+                (max_num_nodes, max_num_nodes, self.num_heads),
+                float('-inf')
+            ).to(device)
+            padded_encoding[0: num_nodes, 0: num_nodes] = dist_embedding
+            spatial_encoding.append(padded_encoding)
+
+        return th.stack(spatial_encoding, dim=0)
+
+
+class SpatialEncoder3d(nn.Module):
+    r"""3D Spatial Encoder, as introduced in
+    `One Transformer Can Understand Both 2D & 3D Molecular Data
+    <https://arxiv.org/pdf/2210.01765.pdf>`__
+    This module encodes pair-wise relation between atoms in the 3D space,
+    by means of Gaussian Basis Kernels and learnable parameters.
+
+    Parameters
+    ----------
+    num_kernels : int
+        Number of Gaussian Basis Kernels to be applied.
+        Each Gaussian Basis Kernel contains learnable kernel center
+        and learnable scaling factor.
+    num_heads : int, optional
+        Number of attention heads if multi-head attention mechanism is applied.
+        Default : 1.
+
+    Examples
+    --------
+    >>> import torch as th
+    >>> import dgl
+    >>> from dgl.nn import SpatialEncoder
+
+    >>> u = th.tensor([0, 0, 0, 1, 1, 2, 3, 3])
+    >>> v = th.tensor([1, 2, 3, 0, 3, 0, 0, 1])
+    >>> g = dgl.graph((u, v))
+    >>> coordinate = th.rand(4, 3)
+    >>> spatial_encoder = SpatialEncoder(4, num_heads=8)
+    >>> out = spatial_encoder(g, coordinate)
+    """
+
+    def __init__(self, num_kernels, num_heads=1):
+        super().__init__()
+        self.num_kernels = num_kernels
+        self.num_heads = num_heads
+        self.embedding_table = nn.Embedding(num_kernels, 2)
+        self.linear_layer_1 = nn.Linear(1, 1)
+        self.linear_layer_2 = nn.Linear(num_kernels, num_kernels)
+        self.linear_layer_3 = nn.Linear(num_kernels, num_heads)

```

Comment:
please. take care of the initialization carefully according to the official implementation